### PR TITLE
Another 6.x support pull request

### PR DESCRIPTION
**Pull request by shribigb**
This is another 6.x pull request. It covers x-pack support and plugin support. Tested installing both 5.6.3 and 6.1.0 version with the same role. I also added adding bootstrap user in case es_api_basic_auth_username is elastic.

### CODE REVIEW

File: tasks/xpack/security/elasticsearch-security.yml [hunk 1]
```diff
@@ -11,6 +11,23 @@
     - es_enable_xpack and '"security" in es_xpack_features'
     - (es_users is defined and es_users.file is defined) or (es_roles is defined and es_roles.file is defined) or (es_role_mapping is defined)
 
+#-----------------------------Create Bootstrap User-----------------------------------
+- name: Check if bootstrap password is set
```
Comment (gregdurham) [inline tasks/xpack/security/elasticsearch-security.yml]: This section causes issues on my setup, when xpack is set to disabled due to the keystore not existing. Should this block of code ever be hit when xpack is not installed? Should we block it with a `when: (es_enable_xpack and '"security" in es_xpack_features')`

Comment (hokeycokey) [reply tasks/xpack/security/elasticsearch-security.yml]: @gregdurham just his this issue too.  You suggestion will work and think its the way to go.

Comment (shribigb) [reply tasks/xpack/security/elasticsearch-security.yml]: Yes I am making those changes, I encountered the same in further testing

Comment (gregdurham) [reply tasks/xpack/security/elasticsearch-security.yml]: Thank you!

File: tasks/java.yml [hunk 2]
```diff
@@ -10,6 +10,12 @@
   yum: name={{ java }} state={{java_state}}
   when: ansible_os_family == 'RedHat'
 
+- name: correct java version selected
```
Comment (gregdurham) [inline tasks/java.yml]: Had issues with this on RHEL 7, the binary java8 does not exist

Comment (gregdurham) [reply tasks/java.yml]: i.e. it is already just java

Comment (shribigb) [reply tasks/java.yml]: Amazon AMIs have different behaviour than RedHat. Made the change and tested on both Amazon AMI and redhat 7.4. Thanks @gregdurham for reviewing

File: tasks/xpack/security/elasticsearch-security.yml (hunk 1 above)
Comment (hokeycokey) [reply tasks/xpack/security/elasticsearch-security.yml]: Thanks!

**Comment by hokeycokey** [comment]
Elasticsearch is failing to start.  Using Vagrant with Centos7 (centos-release-7-4.1708.el7.centos.x86_64).

Get the following error: 

`Jan 11 07:36:38 es-master-data systemd: Starting Elasticsearch-node1...
Jan 11 07:36:38 es-master-data systemd: Failed at step EXEC spawning /usr/share/elasticsearch/bin/elasticsearch-systemd-pre-exec: No such file or directory
Jan 11 07:36:38 es-master-data systemd: node1_elasticsearch.service: control process exited, code=exited status=203
`

**Comment by hokeycokey** [comment]
https://github.com/shribigb/ansible-elasticsearch/blob/3076ed920eacca4f0aa932bea9f9d0b9207a5437/tasks/xpack/security/elasticsearch-security-native.yml#L16

This task fails on the second run if the {{es_api_basic_auth_password}} has been changed from the bootstrap password.

**Comment by shribigb** [comment]
@hokeycokey I will give it a try, from the documentation it seems this should be fine(https://www.elastic.co/guide/en/x-pack/6.1/setting-up-authentication.html#set-built-in-user-passwords).. if you change the password for elastic after setting bootstrap.password it will take the new password. I have a check to find out if bootstrap.password is already set.. in the second run I dont change/set bootstrap password. As long as you are sending new password as part of your playbook run in theory it should work. But I will test it anyways. So to reproduce in first run say you set the password to changeme for elastic user and then changed it to changeme1 through API and in next run of the playbook you are setting es_api_basic_auth_password to changeme1 and it fails there?

**Comment by hokeycokey** [comment]
Hi @shribigb  and thanks.

In defaults I have the following:
```

es_api_basic_auth_username: elastic
es_api_basic_auth_password: changeme
```

Within playbook:

```
vars:
   es_users:
     native:
       elastic: 
         password: changednow
```

Error on second run:

```
ASK [elasticsearch : List Native Users] ***************************************
fatal: [192.168.2.5]: FAILED! => {"changed": false, "content": "{\"error\":{\"root_cause\":[{\"type\":\"security_exception\",\"reason\":\"failed to authenticate user [elastic]\",\"header\":{\"WWW-Authenticate\":\"Basic realm=\\\"security\\\" charset=\\\"UTF-8\\\"\"}}],\"type\":\"security_exception\",\"reason\":\"failed to authenticate user [elastic]\",\"header\":{\"WWW-Authenticate\":\"Basic realm=\\\"security\\\" charset=\\\"UTF-8\\\"\"}},\"status\":401}", "content_length": "345", "content_type": "application/json; charset=UTF-8", "json": {"error": {"header": {"WWW-Authenticate": "Basic realm=\"security\" charset=\"UTF-8\""}, "reason": "failed to authenticate user [elastic]", "root_cause": [{"header": {"WWW-Authenticate": "Basic realm=\"security\" charset=\"UTF-8\""}, "reason": "failed to authenticate user [elastic]", "type": "security_exception"}], "type": "security_exception"}, "status": 401}, "msg": "Status code was not [200]: HTTP Error 401: Unauthorized", "redirected": false, "status": 401, "url": "http://localhost:9200/_xpack/security/user", "www_authenticate": "Basic realm=\"security\" charset=\"UTF-8\""}

```

It seems as though the task uses `es_api_basic_auth_password: changeme`, on every run.  It works on the first run of course.  By the 2nd run the password has changed.

**Comment by shribigb** [comment]
Hi @hokeycokey that's expected behaviour. Role doen't know anything about you have changed password of es_api_basic_auth_username in setting es_users.native.elastic.password to changednow. Rather in first run itself its setting bootstrap.password to changeme and then changing the password of elastic to changednow when its trying to change passwords of native users. In second run in your playbook, you should be considering setting  es_api_basic_auth_password: changednow.

File: tasks/xpack/security/elasticsearch-security.yml (hunk 1 above)
Comment (cvladicgit) [reply tasks/xpack/security/elasticsearch-security.yml]: I am new  to Ansible and Elastic, and I ran the playbook on CentOS 7, i encountered the problem in elasticsearch-security.yml "Ensure x-pack conf directory exists (file)".  Do I have to create the keystore?   My error:

TASK [elasticsearch : Check if bootstrap password is set] ****************************************************************************************************
fatal: [esnode1]: FAILED! => {"changed": true, "cmd": ["/usr/share/elasticsearch/bin/elasticsearch-keystore", "list"], "delta": "0:00:01.361441", "end": "2018-01-23 09:13:37.358360", "failed": true, "msg": "non-zero return code", "rc": 65, "start": "2018-01-23 09:13:35.996919", "stderr": "", "stderr_lines": [], "stdout": "ERROR: Elasticsearch keystore not found. Use 'create' command to create one.", "stdout_lines": ["ERROR: Elasticsearch keystore not found. Use 'create' command to create one."]}

**Comment by cvladicgit** [comment]
I tried this on CentOS 7 to install version 6.1.2 and I am getting an error in elasticsearch-security.yml "Ensure x-pack conf directory exists (file)" task step.  There are some areas where I have to include Become: yes and some with Become_User.  

TASK [elasticsearch : Check if bootstrap password is set] ****************************************************************************************************
fatal: [esnode1]: FAILED! => {"changed": true, "cmd": ["/usr/share/elasticsearch/bin/elasticsearch-keystore", "list"], "delta": "0:00:01.361441", "end": "2018-01-23 09:13:37.358360", "failed": true, "msg": "non-zero return code", "rc": 65, "start": "2018-01-23 09:13:35.996919", "stderr": "", "stderr_lines": [], "stdout": "ERROR: Elasticsearch keystore not found. Use 'create' command to create one.", "stdout_lines": ["ERROR: Elasticsearch keystore not found. Use 'create' command to create one."]}

File: tasks/java.yml [hunk 3]
```diff
@@ -10,6 +10,17 @@
   yum: name={{ java }} state={{java_state}}
   when: ansible_os_family == 'RedHat'
 
+- name: Get the installed java path
+  shell: "update-alternatives --display java | grep '^/' | awk '{print $1}' | grep 1.8.0"
```
[the remark this answers is not in the document]
Comment (shribigb) [reply tasks/java.yml]: This is fixed! I somehow missed a check for RedHat.

**Comment by Crazybus** [comment]
Hi @shribigb 

Thanks for the great work here! I have created a new pull request https://github.com/elastic/ansible-elasticsearch/pull/412 based on your work. I rebased your pull request against master (with the updated `.kitchen.yml` for testing 5.x and 6.x) and I'm now fixing up all of the tests. 

Cheers,

Michael

**Pull request closed without merging**